File: sdks/typescript/src/ui/apply.ts
import { parsePatchEvent, PatchOp, type PatchInput, type PatchEvent } from "./event.ts";
import { cloneJson, hasOwn, isRecord, type JsonRecord } from "./json.ts";

export type Snapshot = JsonRecord;

export function applyPatch(snapshot: Snapshot, patch: PatchInput): Snapshot {
  const event = parsePatchEvent(patch);

  switch (event.op) {
    case PatchOp.START:
      return cloneJson(event.model);
    case PatchOp.SET:
      return applySet(snapshot, event);
    case PatchOp.APPEND:
      return applyAppend(snapshot, event);
    case PatchOp.ERROR:
      return setByMask(snapshot, "meta.error", event.meta);
    case PatchOp.PING:
    case PatchOp.END:
      return snapshot;
  }
}

export function applyPatches(snapshot: Snapshot, patches: Iterable<PatchInput>): Snapshot {
  for (const patch of patches) snapshot = applyPatch(snapshot, patch);
  return snapshot;
}

function applySet(snapshot: Snapshot, event: Extract<PatchEvent, { op: "set" }>): Snapshot {
  if (event.mask === undefined) {
    upsertBlock(snapshot, event.block);
    return snapshot;
  }
  return setByMask(snapshot, event.mask, "meta" in event ? event.meta : event.block);
}

function applyAppend(snapshot: Snapshot, event: Extract<PatchEvent, { op: "append" }>): Snapshot {
  const field = event.mask.slice("block.".length);
  if (!field || field.includes(".")) {
    throw new Error(`unsupported append mask: ${JSON.stringify(event.mask)}; expected 'block.<field>'`);
  }
  if (!event.block.id) throw new Error("append block requires a non-empty id");
  if (!hasOwn(event.block, field)) throw new Error(`append block does not contain field ${JSON.stringify(field)}`);

  const nextValue = event.block[field];
  if (typeof nextValue !== "string" && !Array.isArray(nextValue)) {
    throw new TypeError(`append field ${JSON.stringify(field)} must be a string or list`);
  }

  const blocks = ensureBlocks(snapshot);
  const existing = blocks.find((block) => block.id === event.block.id);
  if (existing) {
    const currentValue = existing[field];
    if (currentValue === undefined || currentValue === null) {
      existing[field] = cloneJson(nextValue);
    } else if (typeof currentValue === "string" && typeof nextValue === "string") {
      existing[field] = currentValue + nextValue;
    } else if (Array.isArray(currentValue) && Array.isArray(nextValue)) {
      existing[field] = [...currentValue, ...cloneJson(nextValue)];
    } else {
      throw new TypeError(`append field ${JSON.stringify(field)} has incompatible value types`);
    }
    return snapshot;
  }

  if (!event.block.type) throw new Error("append for a missing block requires a complete block with type");
  blocks.push(cloneJson(event.block));
  return snapshot;
}

function upsertBlock(snapshot: Snapshot, block: JsonRecord): void {
  if (typeof block.id !== "string" || !block.id) throw new Error("set block requires a non-empty id");
  if (typeof block.type !== "string" || !block.type) throw new Error("set block requires a non-empty type");

  const blocks = ensureBlocks(snapshot);
  const index = blocks.findIndex((current) => current.id === block.id);
  if (index === -1) blocks.push(cloneJson(block));
  else blocks[index] = cloneJson(block);
}

function setByMask(snapshot: Snapshot, mask: string, source: JsonRecord): Snapshot {
  const separator = mask.indexOf(".");
  if (separator <= 0 || separator === mask.length - 1) throw new Error(`invalid set mask: ${JSON.stringify(mask)}`);

  const root = mask.slice(0, separator);
  const path = mask.slice(separator + 1).split(".");
  const value = readPath(source, path);
  if (!value.found) throw new Error(`payload does not contain masked value: ${JSON.stringify(mask)}`);

  let target: JsonRecord;
  if (root === "meta") {
    if (snapshot.meta === undefined) snapshot.meta = {};
    if (!isRecord(snapshot.meta)) throw new TypeError("snapshot meta must be an object");
    target = snapshot.meta;
  } else if (root === "block") {
    const blockId = source.id;
    const block = ensureBlocks(snapshot).find((candidate) => candidate.id === blockId);
    if (!block) throw new Error(`target block does not exist: ${JSON.stringify(blockId)}`);
    target = block;
  } else {
    throw new Error(`unsupported set mask root: ${JSON.stringify(root)}`);
  }

  writePath(target, path, value.value);
  return snapshot;
}

function readPath(source: JsonRecord, path: string[]): { found: boolean; value?: unknown } {
  let current: unknown = source;
  for (const part of path) {
    if (!isRecord(current) || !hasOwn(current, part)) {
      const leaf = path.at(-1);
      return leaf !== undefined && hasOwn(source, leaf)
        ? { found: true, value: source[leaf] }
        : { found: false };
    }
    current = current[part];
  }
  return { found: true, value: current };
}

function writePath(target: JsonRecord, path: string[], value: unknown): void {
  let current = target;
  for (const part of path.slice(0, -1)) {
    if (!isRecord(current[part])) current[part] = {};
    current = current[part] as JsonRecord;
  }
  const leaf = path.at(-1);
  if (leaf === undefined) throw new Error("set path must not be empty");
  current[leaf] = cloneJson(value);
}

function ensureBlocks(snapshot: Snapshot): JsonRecord[] {
  if (snapshot.blocks === undefined) snapshot.blocks = [];
  if (!Array.isArray(snapshot.blocks)) throw new TypeError("snapshot blocks must be an array");
  for (const block of snapshot.blocks) {
    if (!isRecord(block)) throw new TypeError("snapshot block must be an object");
  }
  return snapshot.blocks as JsonRecord[];
}

File: sdks/python/src/agentue/apply.py
"""Deterministic AgentUE reducer used for live updates and replay."""

import copy
from collections.abc import Iterable, Mapping
from typing import Any, TypeAlias

from agentue.event import PatchEvent, PatchOp
from agentue.model import PROTOCOL_VERSION, validate_block

PatchInput: TypeAlias = str | Mapping[str, Any] | PatchEvent
_MISSING = object()


def apply_patch(snapshot: dict[str, Any], patch: PatchInput) -> dict[str, Any]:
    """Apply one event to ``snapshot`` and return the resulting state.

    Incremental operations mutate and return ``snapshot``. ``start`` returns a
    new state because its model replaces the previous state completely.
    Multiplexing callers route by ``stream_id`` first; this reducer handles one model.
    """
    event = _coerce_event(patch)

    if event.op is PatchOp.START:
        return copy.deepcopy(event.model or snapshot)
    if event.op is PatchOp.SET:
        return _apply_set(snapshot, event)
    if event.op is PatchOp.ERROR:
        return _set_by_mask(snapshot, "meta.error", event.meta or {})
    if event.op is PatchOp.APPEND:
        return _apply_append(snapshot, event)

    return snapshot


def apply_patches(snapshot: dict[str, Any], patches: Iterable[PatchInput]) -> dict[str, Any]:
    """Apply an ordered event sequence."""
    for patch in patches:
        snapshot = apply_patch(snapshot, patch)
    return snapshot


def _coerce_event(patch: PatchInput) -> PatchEvent:
    if isinstance(patch, PatchEvent):
        return patch
    if isinstance(patch, str):
        return PatchEvent.model_validate_json(patch)
    return PatchEvent.model_validate(dict(patch))


def _apply_set(snapshot: dict[str, Any], event: PatchEvent) -> dict[str, Any]:
    if event.mask is None:
        _upsert_block(snapshot, event.block or {})
        return snapshot

    source = event.meta if event.meta is not None else event.block
    return _set_by_mask(snapshot, event.mask, source or {})


def _apply_append(snapshot: dict[str, Any], event: PatchEvent) -> dict[str, Any]:
    mask = event.mask or ""
    field = mask.removeprefix("block.")
    if not field or "." in field:
        raise ValueError(f"unsupported append mask: {mask!r}; expected 'block.<field>'")

    block_data = event.block or {}
    block_id = block_data.get("id")
    if not isinstance(block_id, str) or not block_id:
        raise ValueError("append block requires a non-empty id")
    if field not in block_data:
        raise ValueError(f"append block does not contain field {field!r}")

    new_value = block_data[field]
    if not isinstance(new_value, (str, list)):
        raise TypeError(f"append field {field!r} must be a string or list")

    blocks = snapshot.setdefault("blocks", [])
    for existing in blocks:
        if existing.get("id") != block_id:
            continue
        _require_inline_block(existing)
        old_value = existing.get(field)
        if old_value is None:
            existing[field] = copy.deepcopy(new_value)
        elif isinstance(old_value, str) and isinstance(new_value, str):
            existing[field] = old_value + new_value
        elif isinstance(old_value, list) and isinstance(new_value, list):
            existing[field] = old_value + copy.deepcopy(new_value)
        else:
            raise TypeError(f"append field {field!r} has incompatible value types")
        return snapshot

    if not isinstance(block_data.get("type"), str) or not block_data["type"]:
        raise ValueError("append for a missing block requires a complete block with type")
    validate_block(block_data)
    blocks.append(copy.deepcopy(block_data))
    return snapshot


def _upsert_block(snapshot: dict[str, Any], block_data: dict[str, Any]) -> None:
    block_id = block_data.get("id")
    if not isinstance(block_id, str) or not block_id:
        raise ValueError("set block requires a non-empty id")
    validate_block(block_data)
    if "ref" in block_data and snapshot.get("version") != PROTOCOL_VERSION:
        raise ValueError("reference blocks require model version 1.1")

    blocks = snapshot.setdefault("blocks", [])
    for index, existing in enumerate(blocks):
        if existing.get("id") == block_id:
            blocks[index] = copy.deepcopy(block_data)
            return
    blocks.append(copy.deepcopy(block_data))


def _set_by_mask(snapshot: dict[str, Any], mask: str, source: dict[str, Any]) -> dict[str, Any]:
    root, separator, relative_path = mask.partition(".")
    if not separator or not relative_path:
        raise ValueError(f"invalid set mask: {mask!r}")
    path = relative_path.split(".")
    value = _read_path(source, path)
    if value is _MISSING:
        raise ValueError(f"payload does not contain masked value: {mask!r}")

    if root == "meta":
        target = snapshot.setdefault("meta", {})
    elif root == "block":
        block_id = source.get("id")
        target = next((block for block in snapshot.get("blocks", []) if block.get("id") == block_id), None)
        if target is None:
            raise ValueError(f"target block does not exist: {block_id!r}")
        _require_inline_block(target)
    else:
        raise ValueError(f"unsupported set mask root: {root!r}")

    _write_path(target, path, value)
    return snapshot


def _require_inline_block(block: dict[str, Any]) -> None:
    # spec: A pure reducer cannot append to unknown external content.
    if "ref" in block:
        raise ValueError(f"unresolved reference block {block.get('id')!r}; materialize before field patches")


def _read_path(source: dict[str, Any], path: list[str]) -> Any:
    current: Any = source
    for part in path:
        if not isinstance(current, dict) or part not in current:
            # Accept the original flat leaf payload shape for compatibility.
            return source.get(path[-1], _MISSING)
        current = current[part]
    return current


def _write_path(target: dict[str, Any], path: list[str], value: Any) -> None:
    current = target
    for part in path[:-1]:
        child = current.get(part)
        if not isinstance(child, dict):
            child = {}
            current[part] = child
        current = child
    current[path[-1]] = copy.deepcopy(value)
